### PR TITLE
8287818: Shenandoah: adapt nmethod arming from Loom

### PR DESCRIPTION
**Pull request by zhengyu123**
Loom implemented nmethod arming mechanism, which is very similar to Shenandoah's. Shenandoah can use it.

Test:

- [x] hotspot_gc_shenandoah on Linux x86_64 and Windows x64
- [x] tier1 with Shenandoah GC on Linux x86_64 and Windows x64 
- [x] hotspot_gc_shenandoah on AArch64

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287818](https://bugs.openjdk.org/browse/JDK-8287818): Shenandoah: adapt nmethod arming from Loom

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [b4ef6d63](https://git.openjdk.org/jdk/pull/9048/files/b4ef6d638bf943a6f72c14fb3093cf8f3170d638)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9048/head:pull/9048` \
`$ git checkout pull/9048`

Update a local copy of the PR: \
`$ git checkout pull/9048` \
`$ git pull https://git.openjdk.org/jdk pull/9048/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9048`

View PR using the GUI difftool: \
`$ git pr show -t 9048`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9048.diff">https://git.openjdk.org/jdk/pull/9048.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSetNMethod.cpp [hunk 1]
```diff
@@ -69,16 +69,4 @@ bool ShenandoahBarrierSetNMethod::nmethod_entry_barrier(nmethod* nm) {
...
-}
-
-ByteSize ShenandoahBarrierSetNMethod::thread_disarmed_offset() const {
-  return ShenandoahThreadLocalData::disarmed_value_offset();
-}
-
-int* ShenandoahBarrierSetNMethod::disarmed_value_address() const {
-  return ShenandoahCodeRoots::disarmed_value_address();
-}
+}
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahBarrierSetNMethod.cpp:72]: Nit: missing newline.

File: src/hotspot/share/gc/shenandoah/shenandoahCodeRoots.cpp [hunk 2]
```diff
@@ -126,17 +126,7 @@ void ShenandoahCodeRoots::flush_nmethod(nmethod* nm) {
...
-  // 0 is reserved for new nmethod
-  if (_disarmed_value == 0) {
-    _disarmed_value = 1;
-  }
-
-  JavaThreadIteratorWithHandle jtiwh;
-  for (JavaThread *thr = jtiwh.next(); thr != NULL; thr = jtiwh.next()) {
-    ShenandoahThreadLocalData::set_disarmed_value(thr, _disarmed_value);
-  }
+  BarrierSet::barrier_set()->barrier_set_nmethod()->arm_all_nmethods();
```
Comment (shipilev) [inline src/hotspot/share/gc/shenandoah/shenandoahCodeRoots.cpp:129]: Question: in `ShenandoahBarrierSet::on_thread_attach`, we check `barrier_set_nmethod() != NULL`, do we need to do the same check here, or?

**Review by shipilev — COMMENTED** [review]
Looks mostly fine to me, but I have questions.

File: src/hotspot/share/gc/shenandoah/shenandoahBarrierSetNMethod.cpp (hunk 1 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahBarrierSetNMethod.cpp]: Fixed

File: src/hotspot/share/gc/shenandoah/shenandoahCodeRoots.cpp (hunk 2 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahCodeRoots.cpp:129]: Only in `passive` mode, Shenandoah does not install nmethod barrier, but `passive` mode never calls `arm_nmethod()`, etc. nmethod entry barrier related methods.

**Review by shipilev — APPROVED** [review]

**Pull request closed without merging**